modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+      _sanity(_withinBounds(idx));
+
+      if idx == _size then
+        return;
+
+      for i in idx.._size do {
+        ref src = _getRef(i + 1);
+        ref dst = _getRef(i);
+        _move(src, dst);
+      }
+
+      _maybeReleaseMem(1);
+    }
+
+    //
+    // Assumes that a copy of the input element has already been made at some
+    // previous boundary, IE giving append's parameter the "in" intent.
+    //
+    pragma "no doc"
+    proc _append(ref x: eltType) {
+      _maybeAcquireMem(1);
+      ref src = x;
+      ref dst = _getRef(_size + 1);
+      _move(src, dst);
+      _size += 1;
+    }
+
+    /*
+      Add an element to the end of this List.
+
+      :arg x: An element to append.
+    */
+    proc append(pragma "no auto destroy" in x: eltType) {
+      _enter();
+      _append(x);
+      _leave();
+    }
+
+    pragma "no doc"
+    inline proc _extendGeneric(collection) {
+
+      //
+      // TODO: This could be faster if we resized once and then performed
+      // repeated moves, rather than calling _append().
+      //
+      for x in collection do
+        _append(x);
+    }
+
+    //
+    // Can't use this for now without getting warnings, so stick with the
+    // simple extend() implementation.
+    //
+    pragma "no doc"
+    inline proc _extendGenericAlter(ref collection) {
+      const oldSize = _size;
+      
+      _maybeAcquireMem(collection.size);
+      
+      for i in 1..collection.size do {
+        pragma "no auto destroy"
+        var cpy: eltType = collection[i];
+        ref src = cpy;
+        ref dst = _getRef(oldSize + i);
+        _move(src, dst);
+      }
+
+      _size += collection.size;
+    }
+
+    /*
+      Extend this List by appending a copy of each element contained in
+      another List.
+
+      :arg other: A List containing elements of the same type as those
+        contained in this List.
+    */
+    proc extend(other: List(?t)) where t == eltType {
+      _enter();
+      _extendGeneric(other);
+      _leave();
+    }
+
+    /*
+      Extend this List by appending a copy of each element contained in an
+      array.
+
+      :arg other: An array containing elements of the same type as those
+        contained in this List.
+    */
+    proc extend(other: [?d] ?t) where t == eltType {
+      _enter();
+      _extendGeneric(other);
+      _leave();
+    }
+
+    /*
+      Insert an element at a given position in this List, shifting all elements
+      following that index one to the right.
+
+      .. warning::
+      
+        Inserting an element into this List may invalidate existing references
+        to the elements contained in this List.
+
+      :arg i: The index of the element at which to insert.
+      :arg x: The element to insert.
+
+      :throws IllegalArgumentError: If the given index is out of bounds.
+    */
+    proc insert(i: int, pragma "no auto destroy" in x: eltType) throws {
+      _enter();
+      try _boundsCheckLeaveOnThrow(i);
```

Comment:
You need to deinit / destroy element i before calling _expand

Reply:
Why do I need to call deinit when that element got shifted one to the right?

Reply:
Sorry, you are right, you don't.